[SRVCOM-2131] Add section on Serverless upgrades and patches & hotfixes (#64154)
openshift/openshift-docs: modules/serverless-hotfix-patch.adoc

// Module included in the following assemblies:
//
// * install/serverless-upgrades.adoc

:_content-type: CONCEPT
[id="serverless-hotfix-patch_{context}"]
= Serverless operator maintenance release upgrades

== Latest and maintenance releases

Beginning with {ServerlessProductName} 1.29, the different product versions are available as follows:

* The latest release is available through the `stable` channel.

> * The maintenance release is available through its version-based channel, for example, `stable-1.29`. The maintenance release is the release prior to the latest release, for example if `stable` is 1.30, then maintenance release is `stable-1.29`.

Reviewer: ```suggestion
* The maintenance release is available through its version-based channel, for example, `stable-1.29`. The maintenance release is the release prior to the latest release, for example if `stable` channel includes 1.30, then maintenance release will be available in `stable-1.29` channel.
```

Reviewer: It seems like we're calling "stable-1.29" a release. But it is the channel where the release is available. Just a little wording issue.

> Using a version-based channel allows you to stay within a specific `x.y` release. Additionally, it prevents an upgrade to the latest version of the product, which might contain breaking changes.

Reviewer: ```suggestion
Using a version-based channel allows you to stay within a specific `x.y` stream. Additionally, it prevents an upgrade to the latest version of the product, which might contain breaking changes.
```

To switch to the maintenance release, update the channel parameter in the subscription object YAML file from `stable` to the corresponding version-based channel, such as `stable-1.29`.

== Patches and hotfixes for maintenance releases

> Same as the stable release, the maintenance release is subject to patches and hotfixes, which help to maintain your deployment up-to-date with critical bug and security fixes.

Reviewer 2: Suggestions:

```suggestion
As with the stable release, the maintenance release is subject to patches and hotfixes, which help to keep your deployment up to date with critical bug and security fixes.
```

+
Hotfixes might not be available to all customers immediately. However, changes introduced by hotfixes often become available to all customers as part of a future release.
+
When a hotfix pertaining to your deployment is available, you will be given the hotfix CatalogSource to update your subscription and obtain the hotfix.
+
After a new operator release is available, deployed operators with hotfixes can also be upgraded. To use the latest GA version, modify the subscription to use the public CatalogSource instead of the hotfix one.

The following diagram illustrates how patches and hotfixes work:

[source]
----
                                                                            
     stable                                    stable-1.28                  
+--------------+              +--------------------------------------------+
|              |              |                                            |
|  +--------+  | upgrades to  |   +--------+    +--------+    +--------+   |

> |  | 1.28.0 |-------------------> | 1.28.0 |    | 1.28.1 |    | 1.28.2 |   |

Reviewer: I'm curious if this horizontal line with "upgrades to" is correct. The user doesn't ever upgrade from 1.28.0 to 1.28.0. What is this line supposed to depict? Switching channels?

Author: @mgencur Good point! I've changed it from `upgrades to` to `corresponds to`.

|  +--------+  |                created|                   |upgrades    |      
|  | 1.28.1 |  |                from   |     hotfix_xyz    |to          |      
|  +--------+  |                       |   +------------+  |            |      
|              |                       +-->|            |--+            |      
|              |                           |            |               |      
|  +--------+  | upgrades to               +------------+               |      
|  | 1.29.0 |<----------------------------------------------------------+      
|  +--------+  |                                                               
|              |                                                               
|              |                                                               
|  +--------+  |                                                               
|  | 1.30.0 |  |                                                               
|  +--------+  |                                                               
|              |                                                               
+--------------+                                                               

> ----

Reviewer 2: Unfortunately, I don't think these sort of ascii diagrams are allowed in our docs. You'll have to request that they are made into formal diagrams through our team's graphic designer. You can submit a request through this form: [Technical Diagram Request Form](https://docs.google.com/forms/d/e/1FAIpQLSdF8V0n3E-aAfRen8vZchXCxwA0iDTd0QYyASnutznfGLATvA/viewform?usp=sf_link)

If you need to merge this PR before the diagrams can be made, I would recommend commenting them out pre-merge, so they're still in this file for reference. Let me know if you have any questions about this process!

== Upgrade path for maintenance releases

> If you use a version-based channel, you can always upgrade to the latest version in the channel, or head, for example, from 1.29.0 to 1.29.2 in the `stable-1.29` channel.

Reviewer 2: Suggestion to break up this sentence and avoid all the commas:

```suggestion
If you use a version-based channel, you can always upgrade to the latest version in the channel, or head. For example, you can upgrade from 1.29.0 to 1.29.2 in the `stable-1.29` channel.
```

Additionally, from the head of the channel, you can upgrade to the next `x.y` release. For example, if 1.29.2 is the head in the `stable-1.29` channel, you can upgrade from 1.29.2 to 1.30. Such cross-channel updates are not done automatically, and the administrator needs to switch the channel manually by updating the subscription.

== Upgrade examples

=== Scenario 1

> In this scenario:

Reviewer 2: The [ISG](https://www.ibm.com/docs/en/ibm-style?topic=format-lists#lead-in-wording) recommends we avoid list introductions that do not form a complete sentence: "To avoid possible translation problems, use complete sentences to introduce lists and not phrases". Here's one way to work around that:

```suggestion
In this scenario, the following circumstances are true:
```

> * the channel is `stable-1.28`
> * you are switching to `stable`
> * the currently installed version is 1.28.0

Reviewer 2: Per the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=format-lists#capitalization-in-lists), the first word in each list item needs to be capitalized (plus one grammar nit):

```suggestion
* The channel is `stable-1.28`
* You are switching to the `stable` channel
* The currently installed version is 1.28.0
```

> * 1.29.0 was released earlier than 1.28.1

Reviewer 2: Nit:

```suggestion
* 1.29.0 was released before 1.28.1
```

* 1.30.0 is the head of `stable`

The upgrade path from 1.28.0 on `stable-1.28` to 1.29.0 on `stable` in this scenario is 1.28.0 to 1.28.1 to 1.29.0.

[source]
----
                                            
     stable                    stable-1.28  
+--------------+            +--------------+
|              |            |              |
|  +--------+  |            |  +--------+  |
|  | 1.28.0 |  |            |  | 1.28.0 |  |
|  +--------+  |            |  +--------+  |
|              |            |       |      |
|              |            |       |      |
|  +--------+  |            |       |      |
|  | 1.29.0 |<--------      |       v      |
|  +--------+  |     |      |  +--------+  |
|              |     +---------| 1.28.1 |  |
|              |            |  +--------+  |
|  +--------+  |            |              |
|  | 1.30.0 |  |            |              |
|  +--------+  |            |              |
|              |            |              |
+--------------+            +--------------+

> ----

Reviewer 2: Same feedback regarding my comment above about requesting a diagram and excluding this from the PR in the meantime.

=== Scenario 2

> In this scenario:

Reviewer 2: Same as above:

```suggestion
In this scenario, the following circumstances are true:
```

> * the channel is `stable-1.29`
> * the currently installed version is 1.29.0
> * 1.29.1 was released to both `stable-1.29` and `stable` before `1.30.0` was released to `stable`

Reviewer 2: Nits regarding item capitalization and using "channel" more explicitly:

```suggestion
* The channel is `stable-1.29`
* The currently installed version is 1.29.0
* 1.29.1 was released to both the `stable-1.29` and `stable` channels before `1.30.0` was released to the `stable` channel
```

The upgrade path from 1.29.0 on `stable-1.29` to 1.30.0 on `stable` in this scenario is 1.29.0 to 1.29.1 to 1.30.0.

[source]
----
                                           
     stable                   stable-1.29  
+--------------+           +--------------+
|              |           |              |
|  +--------+  |           |  +--------+  |
|  | 1.29.0 |  |           |  | 1.29.0 |  |
|  +--------+  |           |  +--------+  |
|              |           |       |      |
|              |           |       v      |
|  +--------+  |           |  +--------+  |
|  | 1.29.1 |  |           |  | 1.29.1 |  |
|  +--------+  |           |  +--------+  |
|              |           |       |      |
|              |           |       |      |
|  +--------+  |           |       |      |
|  | 1.30.0 |<---------------------+      |
|  +--------+  |           |              |
|              |           |              |
+--------------+           +--------------+

> ----

Reviewer 2: Same feedback regarding my comment above about requesting a diagram and excluding this from the PR in the meantime.

=== Scenario 3

> In this scenario:

Reviewer 2: ```suggestion
In this scenario, the following circumstances are true:
```

Using a version-based channel allows you to stay within a specific `x.y` release. Additionally, it prevents an upgrade to the latest version of the product, which might contain breaking changes.

To switch to the maintenance release, update the channel parameter in the subscription object YAML file from `stable` to the corresponding version-based channel, such as `stable-1.29`.

== Patches and hotfixes for maintenance releases

Same as the stable release, the maintenance release is subject to patches and hotfixes, which help to maintain your deployment up-to-date with critical bug and security fixes.

* Patches are updates that are distributed as z-releases, for example, {ServerlessProductName} 1.29.1 is the patch that provides updates made since version 1.29.0.
* Hotfixes are fixes that require zero downtime and are used directly in production. They are different from the usual updates in that they upgrade the customer's deployed version, and not the latest released version.
+
Hotfixes might not be available to all customers immediately. However, changes introduced by hotfixes often become available to all customers as part of a future release.
+

> When a hotfix pertaining to your deployment is available, you will be given the hotfix CatalogSource to update your subscription and obtain the hotfix.

Reviewer 3: Is there a link about this for the customer to go from here?

Author: @skonto According to the hotfix process document, specific customers are contacted by the productization team specifically:
"So after the build and distribution (errata publication) its on Engineering/Software Production to tell CEE or the requester of the hotfix, what the CatalogSource image is so that they can communicate to the customer/administrator how to consume the hotfix."

+

> After a new operator release is available, hotfix operators deployed in clusters can also be upgraded. To use the latest GA version, modify the subscription to use the public CatalogSource instead of the hotfix one.

Reviewer 3: Maybe simplify: "...available,already deployed operators with hotfixes can be upgraded"
